### DISCUSSION TITLE
Overloaded generic constraints

### DISCUSSION THREAD

**Discussion by ghost** [dotnet/csharplang#5134] [General]
#### Example
```
    public unsafe struct Pointer<TUnmanaged> where TUnmanaged : unmanaged
    {
        public Pointer(TUnmanaged* ptr)
        {
            this.ptr = ptr;
        }

        private readonly void* ptr;
    }

    public unsafe struct Pointer<TDelegate> where TDelegate : Delegate
    {
        public Pointer(TDelegate method)
        {
            ptr = method != null ? (void*)Marshal.GetFunctionPointerForDelegate(method) : null;
        }

        private readonly void* ptr;
    }
```

**Comment by HaloFour** [comment]
Generic type constraints aren't a part of a type/method signature in the CLR.  Changing that at this point would be a pretty big breaking change.

**Comment by hez2010** [comment]
This becomes more and more useful especially since C# introduced SIMs. Although this would require a CLR change, in my opinion it worth because it will enable sum types for generics, and also this has to be done sooner or later if C# plans to improve generics.
